src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +336,135 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
         }
 
+        /// <summary> A synthesized entrypoint that forwards all calls to an async Main Method </summary>
+        internal sealed class AsyncForwardEntryPoint : SynthesizedEntryPointSymbol
+        {
+            /// The user-defined asynchronous main method.
+            private readonly MethodSymbol _userMain;
+
+            private readonly MethodSymbol _getAwaiterMethod;
+            private readonly MethodSymbol _getResultMethod;
+
+            private readonly ImmutableArray<ParameterSymbol> _parameters;
+
+            /// <summary>
+            /// Task -> Void
+            /// Generic Task -> Int32
+            /// </summary>
+            private static TypeSymbol TranslateReturnType(CSharpCompilation compilation, TypeSymbol returnType, DiagnosticBag diagnosticBag, MethodSymbol userMain)
+            {
+                var specialType = returnType.IsGenericTaskType(compilation) ? SpecialType.System_Int32 : SpecialType.System_Void;
+                var syntax = ExtractReturnTypeSyntax(userMain.DeclaringSyntaxReferences);
+                return Binder.GetSpecialType(compilation, specialType, syntax, diagnosticBag);
+            }
+
+            private static SyntaxNode ExtractReturnTypeSyntax(IEnumerable<SyntaxReference> userMainDeclaringReferences) {
+                var methodSyntax = userMainDeclaringReferences
+                    .Select(r => r.GetSyntax() as MethodDeclarationSyntax)
+                    .Where(s => (object)s != null).FirstOrDefault();
+                return methodSyntax?.ReturnType;
+            }
+
+            internal AsyncForwardEntryPoint(CSharpCompilation compilation, DiagnosticBag diagnosticBag, NamedTypeSymbol containingType, MethodSymbol userMain) :
+                base(containingType, TranslateReturnType(compilation, userMain.ReturnType, diagnosticBag, userMain))
+            {
+                // There should be no way for a userMain to be passed in unless it already passed the 
+                // parameter checks for determining entrypoint validity.
+                Debug.Assert(userMain.ParameterCount == 0 || userMain.ParameterCount == 1);
+
+                _userMain = userMain;
+                _parameters = SynthesizedParameterSymbol.DeriveParameters(userMain, this);
+
+                var userMainReturnTypeSyntax = ExtractReturnTypeSyntax(userMain.DeclaringSyntaxReferences);
+                _getAwaiterMethod = GetRequiredMethod(_userMain.ReturnType, WellKnownMemberNames.GetAwaiter, diagnosticBag, userMainReturnTypeSyntax.Location);
+                if ((object)_getAwaiterMethod != null)
+                {
+                    _getResultMethod = GetRequiredMethod(_getAwaiterMethod.ReturnType, WellKnownMemberNames.GetResult, diagnosticBag, userMainReturnTypeSyntax.Location);
+                    if ((object)_getResultMethod != null && _getResultMethod.ReturnType != ReturnType)
+                    {
+                        // Prototype: This needs a diagnostic
```

Comment:
>Prototype: [](start = 27, length = 10)

The guideline is to use "PROTOTYPE(<feature name>):" at the beginning of comments like this.